File: Stringy.hs
module Stringy
( interpretStringy
, interpret
, intr
) where

import Data.Bits
import Data.Char
import Control.Monad (join)
import Data.List (elemIndex)

-- map function over a string 
applyTo :: (Int -> Int) -> String -> String
applyTo f = map (chr . (`mod` 128) . f . ord)

-----------------------------------------------------
-- The following is a list of the Stingy functions --
-- (and some helper functions as well)             --
-- determined by the Char in the string.           --
-- They are kept in a master list of functions to  --
-- be called as needed.                            --
-- These have the form:                            --
-- (String -> Char -> String -> String)            --
-----------------------------------------------------

-- the general forms these functions take
-- used to reduce typing later on and show functions in point-free form
form0 :: (Int -> Int) -> String -> Char -> String -> String
form0 f xs x ys = (applyTo f xs)++[x]++ys
form1 :: (Int -> Int) -> String -> Char -> String -> String
form1 f xs x ys = (applyTo f xs)++ys
form2 :: (Int -> Int -> Int) -> String -> Char -> String -> String
form2 f xs x ys = form1 (f $ ord $ head ys) xs x (tail ys)
form3 :: (String -> String) -> String -> Char -> String -> String
form3 f xs x ys = (f xs)++ys

-- The identity function
id' = form0 id
-- The increment function
inc = form1 succ
-- The decrement function
dec = form1 pred
-- The shift left function
(<.) :: Bits a => a -> a
(<.) = flip shiftL 1
shiftSL = form1 (<.)
-- The shift right function
(.>) :: Bits a => a -> a
(.>) = flip shiftR 1
shiftSR = form1 (.>)
-- The append digit function
appendDigit :: Int -> String -> Char -> String -> String
appendDigit n = form1 ((+n) . (*10))
-- The modulo function
mod' = form2 (flip mod)
-- The addition function
plus = form2 (+)
-- The subtraction function
minus = form2 subtract
-- The multiplication function
mult = form2 (*)
-- the divistion function
divi = form2 (flip div)
-- The repeat fucntion
rep xs x ys = (join $ replicate (ord $ head ys) xs)++(tail ys)
-- The pivot function
piv xs x ys = (init xs)++[head ys]++[last xs]++(tail ys)
-- The throw forward function
front xs x ys = [head ys]++xs++(tail ys)
-- The throw back function
back xs x ys = xs++(tail ys)++[head ys]
-- The reverse function
rev = form3 reverse
-- The shuffle function
shuff' :: String -> String
shuff' xs
    | odd n = (zipple (splitAt (s+1) xs))++[x']
    | otherwise = zipple (splitAt s xs)
    where n = length xs
          s = n `div` 2
          x' = last . fst $ splitAt (s + 1) xs
          zipple (p,q) = foldl (\acc (x,y) -> acc++[x,y]) "" $ zip p q

shuff = form3 shuff'
-- The greater than or equal function
grteq xs x ys = (init xs)++(if (last xs) > (head ys) then "1" else "0")++(tail ys)
-- The if-then statement
ifst' xs x ys
    | odd . ord . last $ xs = (init xs)++(removeBrace ys)
    | otherwise = (init xs)++(tail $ dropWhile (/= '}') ys)
    where removeBrace = (\st -> (takeWhile (/= '}') st)++(tail $ dropWhile (/= '}') st))
ifst xs x ys = if (elem '}' ys) then ifst' xs x ys else (if odd . ord $ last xs then (init xs)++ys else (init xs)++(tail ys))
-- The delete function
del = form3 (\x -> "")
-- Get string function (inputs '\500' - special character)
getString = form3 (\x -> x++['\500']) 
-- Subroutine function calls intr below
subr xs x ys = (fst $ intr 0 (xs,""))++ys
-- Jump to function
jump xs x ys = form3 id xs x (tail ys)
-- Equal function
eq xs x ys = (init xs)++(if (last xs) == (head ys) then "1" else "0")++(tail ys)
-- length of previous string
length' = form3 (\x -> x++[chr . length $ x])
-- Keeps the preceding string as a Subroutine
keep = del
-- call the kept function
call ext = form3 (\x -> x++ext)
-- Uses up the current char but changes nothing otherwise
use = form1 id
-- The terminate funcion
-- Same as use just another name for clarity
end = use

---------------------------------------------------------
-- hand (for handle) calls the string function based   --
-- on the current character selected. Originally was a --
-- list of functions here using guards                 --
-- Some of this behavior is undefined yet - 09/06/18   --
---------------------------------------------------------
hand :: Char -> (String -> Char -> String -> String)
hand x
    | x == ' ' = getString -- read string input handled elsewhere
    | x == '!' = use -- back skip function handled elsewhere
    | x == '\"' = id' -- undefined
    | x == '#' = length' -- length
    | x == '$' = use -- skip function handled elsewhere
    | x == '%' = mod'
    | x == '&' = id' -- undefined
    | x == '\'' = id' -- undefined
    | x == '(' = front
    | x == ')' = back
    | x == '*' = mult
    | x == '+' = plus
    | x == ',' = use -- set command pointer to 0, handled elsewhere
    | x == '-' = minus
    | x == '.' = end
    | x == '/' = divi
    | isDigit x = (appendDigit $ read [x])
    | x == ':' = keep -- save routine
    | x == ';' = subr 
    | x == '<' = shiftSL
    | x == '=' = eq -- equal
    | x == '>' = shiftSR
    | x == '?' = shuff
    | x == '@' = jump -- jump function handled elsewhere
    | isLetter x = id' -- Letters are non-ops 
    | x == '[' = inc
    | x == '\\' = grteq
    | x == ']' = dec
    | x == '^' = piv
    | x == '_' = rep
    | x == '`' = id' -- undefined
    | x == '{' = ifst
    | x == '|' = call -- call function
    | x == '}' = id' -- part of the ifst
    | x == '~' = rev
    | x == '\DEL' = del
    | otherwise = id' -- ASCII 0 thru 31 non ops 

-- Changes command pointer based on command
determinePointer :: Char -> Int -> Int
determinePointer x n
    | ((isLetter x) || (x == '(') || (x == '$') || (x == '^') || (x < ' ')) = (n + 1)
    | (x == '{') = (n - 1)
    | (x == '!') = (n - 2)
    | (x == ',') || (x =='\DEL') = 0
    | otherwise = n

---------------------------------------------
---------------------------------------------
-- intr is for manipulating the string     --
-- execute the command at the nth char     --
-- and continue this recursively           --
-- essentially handles all non-IO commands --
---------------------------------------------
---------------------------------------------

intr :: Int -> (String,String) -> (String,String)
intr n (prog,ext)
    | prog == "" = error "" -- error on empty string
    | (n >= length prog) || (n < 0) = intr (n `mod` (length prog)) (prog,ext) 
    | (x == '.') || (x == ' ') = (newprog,ext) -- done no call to intr
    | (x == ':') = intr m (newprog,xs)
    | otherwise = intr m (newprog,ext) -- all other commands handled here
    where x = (prog!!n) -- char at current command pointer
          xs = fst $ splitAt n prog -- string before at current command
          ys = tail . snd $ splitAt n prog -- string after current command
          newprog = (hand x) xs x ys -- string result after current command
          m = if (x == '_') || (x == ';') || (x=='|') then (length newprog) - (length ys) else (if x == '@' then ord $ head ys else determinePointer x n) -- special case for _ otherwise use determinePointer

------------------------
-- Run time functions --
------------------------

-- Contstant to indicate where input is needed
inputMarker :: Char
inputMarker = '\500'

-- Constant as string for starting a program
-- (considering intial program as user input)
start :: ([Char],String)
start = (inputMarker:[],"")

-- special index -1 if not found
elemIndex' :: (Eq a) => a -> [a] -> Int
elemIndex' x xs = case (elemIndex x xs) of
                      Just n -> n
                      Nothing -> -1

-- replace nth char with new string
inject :: Int -> String -> String -> String
inject n org inp = xs++inp++ys
    where z = splitAt n org
          xs = fst z
          ys = tail . snd $ z

-- pointer is the index of special \500 char 
-- this is -1 if not located
-- If pointer is -1 then return the program
-- newpointer is the length of input plus current pointer
-- newprogram we replace \500 with the input
-- run intr on newprogram at newpointer
-- pass this to interpretStringy
interpretStringy :: (String,String) -> IO (String,String)
interpretStringy program = do
    let pointer = elemIndex' inputMarker (fst program)
    if pointer < 0 
        then return program 
        else do
            input <- getLine
            let newpointer = (length input) + pointer
                newprogram = (inject pointer (fst program) input, snd program)
            interpretStringy $ intr newpointer newprogram

-- pass start to interpretStringy
-- bind this result
interpret = do
    (result,ext) <- interpretStringy start
    putStrLn result

-- $~[!,| :[+$|$+${$z;["| d( (r( (k( (`( (E( (-(~(z:-#+#(#'#+# #'#]#<#'#+###'#Z#'#]#'#.!)^0-0$H@~$!.!T!r!u!e!
-- ^ This should be a valid program for palindrome detection

File: stringy.py
from collections import defaultdict


dev_mode = False
subroutine = ''

def params(st,pt):
  pt = pt % len(st)
  s1 = st[:pt]
  c = st[pt]
  s2 = st[(pt+1):]
  return (c,s1,s2)

def com_f(res,npt):
  if dev_mode:
    print('Pointer at {npt}: ' + res)
  ps = params(res,npt)
  return hand(npt,*ps)

def help_f(f,s1):
  temp_m = map(ord,s1)
  temp_m = map(f,temp_m)
  temp_m = map(lambda x: x % 128,temp_m)
  temp_m = map(chr,temp_m)
  return ''.join(list(temp_m))

def idty(s1,s2,pt):
  res = s1+s2
  return com_f(res,pt)

def read(s1, s2, pt):
  inp = input()
  res = s1+inp+s2
  pt = len(s1 + inp)
  return com_f(res,pt)

def back_jump(s1,s2,pt):
  res = s1 + s2
  return com_f(res,pt-2)

def length(s1,s2,pt):
  res = s1 + chr(len(s1)) + s2
  return com_f(res,pt+1)

def skip(s1,s2,pt):
  res = s1 + s2
  return com_f(res,pt+1)

def mod(s1,s2,pt):
  modulo = ord(s2[0])
  res = help_f(lambda x: x % modulo,s1) + s2[1:]
  return com_f(res,pt)

def front(s1,s2,pt):
  res = s2[0] + s1 + s2[1:]
  return com_f(res,pt+1)

def back(s1,s2,pt):
  res = s1 + s2[1:] + s2[0]
  return com_f(res,pt)

def op1(s1,s2,pt):
  return

def plus(s1,s2,pt):
  n = ord(s2[0])
  res = help_f(lambda x: x + n, s1) + s2[1:]
  return com_f(res,pt)

def ret(s1,s2,pt):
  res = s1 + s2
  return com_f(res,0)

def minus(s1,s2,pt):
  n = ord(s2[0])
  res = help_f(lambda x: x - n, s1) + s2[1:]
  return com_f(res,pt)

def divi(s1,s2,pt):
  n = ord(s2[0])
  res = help_f(lambda x: x // n, s1) + s2[1:]
  return com_f(res,pt)

def appendDig(d):
  def r_func(s1, s2, pt):
    res = help_f(lambda x: 10*x + d,s1) + s2
    return com_f(res,pt)
  return r_func

def keep(s1,s2,pt):
  subroutine = s1
  return com_f(s2,0)

def subr(s1,s2,pt):
  res = hand(0,*params(s1,0)) + s2
  pt = len(res) - len(s2)
  return com_f(res,pt)

def shiftL(s1,s2,pt):
  res = help_f(lambda x: x << 1,s1) + s2
  return com_f(res,pt)

def eq(s1,s2,pt):
  res = s1[:-1] + str(int(s1[-1] == s2[0])) + s2[1:]
  return com_f(res,pt)

def shiftR(s1,s2,pt):
  res = help_f(lambda x: x >> 1,s1) + s2
  return com_f(res,pt)

def shuff():
  return

def jump(s1,s2,pt):
  res = s1 + s2[1:]
  pt = ord(s2[0])
  return com_f(res,pt)

def inc(s1,s2,pt):
  res = help_f(lambda x: x + 1, s2) + s2
  return com_f(res,pt)

def grteq(s1,s2,pt):
  res = s1[:-1] + str(int(s1[-1] > s2[0])) + s2[1:]
  return com_f(res,pt)

def dec(s1,s2,pt):
  res = help_f(lambda x: x - 1, s2) + s2
  return com_f(res,pt)

def piv(s1,s2,pt):
  res = s1[:-1] + s2[0] + s1[-1] + s2[1:]
  return com_f(res,pt+1)

def rep(s1,s2,pt):
  res = (s1 * (ord(s2[0]))) + s2[1:]
  pt = len(s2) * ord(s2[0])
  return com_f(res,pt)

def ifst():
  return

def call(s1,s2,pt):
  res = s1 + subroutine + s2
  pt = len(s1 + subroutine)
  return com_f(res,pt)

def rev(s1,s2,pt):
  res = s1[::-1] + s2
  return com_f(res,pt)

def delete(s1,s2,pt):
  return com_f(s2,0)

def end(s1,s2,pt):
  return s1+s2

coms = defaultdict(idty,{
    ' ' : read,
    '!' : back_jump,
    '"' : idty,
    '#' : length,
    '$' : skip,
    '%' : mod,
    '&' : idty,
    '\'' : idty,
    '(' : front,
    ')' : back,
    '*' : op1,
    '+' : plus,
    ',' : ret,
    '-' : minus,
    '.' : end,
    '/' : divi,
    '0' : appendDig(0),
    '1' : appendDig(1),
    '2' : appendDig(2),
    '3' : appendDig(3),
    '4' : appendDig(4),
    '5' : appendDig(5),
    '6' : appendDig(6),
    '7' : appendDig(7),
    '8' : appendDig(8),
    '9' : appendDig(9),
    ':' : keep,
    ';' : subr,
    '<' : shiftL,
    '=' : eq,
    '>' : shiftR,
    '?' : shuff,
    '@' : jump,
    '[' : inc,
    '\\' : grteq,
    ']' : dec,
    '^' : piv,
    '_' : rep,
    '`' : idty,
    '{' : ifst,
    '|' : call,
    '}' : idty,
    '~' : rev,
    '\DEL' : delete
})


def hand(pt,c,s1,s2):
  com_num = ord(c) - 32
  command = coms[com_num]
  return command(s1,s2,pt)


def run_stringy(st):
  print(hand(0,*params(st,0)))

run_stringy(input())

